lang/c++/impl/CustomAttributes.cc

```
@@ -16,26 +16,63 @@
+// handle/implement it.
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#endif
+
 #include <map>
 #include <memory>
 
+#include "CustomAttributes.hh"
+#include "Exception.hh"
+
+#include "json/JsonDom.hh"
+
 namespace avro {
 
+CustomAttributes::CustomAttributes(ValueMode valueMode) {
+    switch (valueMode) {
+    case ValueMode::STRING:
+    case ValueMode::JSON:
+        valueMode_ = valueMode;
+        break;
+    default:
+        throw Exception("invalid ValueMode: " + std::to_string(static_cast<int>(valueMode)));
+    }
+}
+
 std::optional<std::string> CustomAttributes::getAttribute(const std::string &name) const {
-    std::optional<std::string> result;
-    std::map<std::string, std::string>::const_iterator iter =
-        attributes_.find(name);
+    auto iter = attributes_.find(name);
     if (iter == attributes_.end()) {
-        return result;
+        return {};
     }
-    result = iter->second;
-    return result;
+    return iter->second;
 }
 
 void CustomAttributes::addAttribute(const std::string &name,
                                     const std::string &value) {
+    // Validate the incoming value.
+    //
+    // NOTE: This is a bit annoying that we accept the data as a string instead of
+    // as an Entity. That means the compiler must convert the value to a string only
+    // for this method to convert it back. But we can't directly refer to the
+    // json::Entity type in the signatures for this class (and thus cannot accept
+    // that type directly as a parameter) because then it would need to be included
+    // from a header file: CustomAttributes.hh. But the json header files are not
+    // part of the Avro distribution (intentionally), so CustomAttributes.hh cannot
+    // #include any of the json header files.
+    const std::string &jsonVal = (valueMode_ == ValueMode::STRING)
+                ? std::move("\"" + value + "\"")
+                : value;
+    try {
```

Is it expensive to validate the input? If we don't do this, what will happen for malformed input?

---

If we don't validate it, then we could end up generating invalid Avro files.

I suppose we could also make a private method that skips the validation and declare the Compiler as a friend class that can invoke it. From the compiler, the values will have already been validated when the JSON schema was originally parsed, so it would be fine to skip in those cases.

The existing "string only" usage is error-prone enough, since it still requires that interior quotes and newlines be escaped (since all it does is add a leading and trailing `"` to turn it into a JSON value). But I figured it would be even more error-prone with allowing any arbitrary JSON element (including complicated values like arrays and objects) and also more error-prone now that it has a modality that is defined at construction time. Failing fast seemed like a much better developer experience than ignoring such issues and letting the library generate an invalid file.

---

> I suppose we could also make a private method that skips the validation and declare the Compiler as a friend class that can invoke it.

Welp, this won't work. The compiler isn't a class but a set of functions. But the real issue is that the `CustomAttributes.addAttribute` method is called from a static function in `Compiler.cc, so we can't possibly make it a friend of `CustomAttributes`. If you have any other recommendation for how we might omit the validation just from the compiler, I will happily implement it.

I suppose another option is to just make it another constructor argument, like `ValidationMode`, and then document that it's unwise to use the "no validate" mode unless the caller is doing something else to validate the values before providing them, since it could otherwise result in invalid/unreadable Avro files.

What do you think?

---

That may introduce a larger change so I think we need opinions from the maintainers. Let's keep it simple in this PR.

---

The other approach is to simply remove validation entirely. But I feel bad doing that seems it makes the API a foot-gun: it becomes easy to accidentally produce an Avro file that is junk/unreadable, even though the function calls that produced the file never returned any errors or threw any exceptions.

FWIW, in the _expected_ cases, where custom values are not present in great number and have values that are simple scalars and short strings, it should be very cheap relative to the rest of the process of encoding and creating the Avro file. For the compiler, reading and processing an Avro file, the extra overhead would be strictly less than the processing needed to initially parse the JSON schema (likely _much_ less, since custom attributes are typically a very small portion of the JSON). And I suspect (though have no profile data) that parsing the JSON schema is already a fairly light-weight part of reading an Avro file.

@martin-g, thoughts?